### DISCUSSION TITLE
How to access request headers in NextJs 13 (app dir based)?

### DISCUSSION THREAD

**Discussion by vikram-vn** [vercel/next.js#49869] [Help]
### Summary

I want to access some of the request headers, and I tried to access them using the below code, but I ended up with no luck. Can anyone say what is wrong with this? I tried to find a solution, but again, I didn't get any answers.

```javascript
const requestHeaders = new Headers(request.headers);
const accessTokenFromHeader = requestHeaders.get('x-access-token');
```

**Comment by ilhamgepe** [comment]
i use nextjs v13.4 this code work for me
docs: https://nextjs.org/docs/app/api-reference/functions/headers#headers
```
import { headers } from "next/headers";

export async function POST(req: Request) {
  const headersList = headers();
  const referer = headersList.get("authorization");

  console.log({ headersList, referer });

  return new Response(JSON.stringify({ message: "test" }), {
    status: 200,
    headers: {
      "Content-Type": "application/json",
    },
  });
}

**Reply by abolfazlakbarzadeh** [reply]
hi, and if we use 'use client' ?

**Reply by hamlim** [reply]
You'll need to wrap your client component in a server component that extracts the header value and then can pass it into the client component

**Reply by abolfazlakbarzadeh** [reply]
yes, you are right, but may we fall into a circle, in this case, we can use a context wrapper, I think the pages/ approach is more comfortable than app/, there are too many issues with 13

**Reply by salman-ryo** [reply]
Thank you for this great solution. Even chatGPT or stack overflow couldn't help me with this, but your answer did the job. 😭🙌

**Reply by djanowski** [reply]
ChatGPT isn't new enough to know about Next.js 13 :)

**Reply by salman-ryo** [reply]
I thought it would be backward compatible and that it will work in case I use the same method as NextJS 12 to get the headers :')

**Reply by Venediktoff** [reply]
You want to use phind.com if you want AI tailored for devs. You can also provide it with a link to documentation or a website.

**Comment by DedaDev** [comment]
any idea how to read in pages?
```ts
context.res.getHeader('prop')
```
doesn't seem to work anymore

**Reply by rimzzlabs** [reply]
inside `getServerSideProps` try

```ts
export async function getServerSideProps(ctx){
 let someHeader = ctx.req.headers['my-custom-headers']
}
```
